homeassistant/components/mqtt/discovery.py

```diff
@@ -155,116 +315,129 @@ def async_discovery_message_received(msg: ReceiveMessage) -> None:  # noqa: C901
-                        if topic[-1] == TOPIC_BASE:
-                            availability_conf[CONF_TOPIC] = f"{topic[:-1]}{base}"
-
-        # If present, the node_id will be included in the discovered object id
-        discovery_id = f"{node_id} {object_id}" if node_id else object_id
-        discovery_hash = (component, discovery_id)
+        for component_config in discovered_components:
+            component = component_config.component
+            node_id = component_config.node_id
+            object_id = component_config.object_id
+            discovery_payload = component_config.discovery_payload
+            if component not in SUPPORTED_COMPONENTS:
+                _LOGGER.warning("Integration %s is not supported", component)
+                return
 
-        if discovery_payload:
-            # Attach MQTT topic to the payload, used for debug prints
-            setattr(
-                discovery_payload,
-                "__configuration_source__",
-                f"MQTT (topic: '{topic}')",
-            )
-            discovery_data = {
-                ATTR_DISCOVERY_HASH: discovery_hash,
-                ATTR_DISCOVERY_PAYLOAD: discovery_payload,
-                ATTR_DISCOVERY_TOPIC: topic,
-            }
-            setattr(discovery_payload, "discovery_data", discovery_data)
+            if TOPIC_BASE in discovery_payload:
+                _replace_topic_base(discovery_payload)
 
-            discovery_payload[CONF_PLATFORM] = "mqtt"
+            # If present, the node_id will be included in the discovery_id.
+            discovery_id = f"{node_id} {object_id}" if node_id else object_id
+            discovery_hash = (component, discovery_id)
 
-        if discovery_hash in mqtt_data.discovery_pending_discovered:
-            pending = mqtt_data.discovery_pending_discovered[discovery_hash]["pending"]
-            pending.appendleft(discovery_payload)
-            _LOGGER.debug(
-                "Component has already been discovered: %s %s, queuing update",
-                component,
-                discovery_id,
-            )
-            return
+            if discovery_payload:
+                # Attach MQTT topic to the payload, used for debug prints
+                setattr(
+                    discovery_payload,
+                    "__configuration_source__",
+                    f"MQTT (topic: '{topic}')",
+                )
+                discovery_data = {
+                    ATTR_DISCOVERY_HASH: discovery_hash,
+                    ATTR_DISCOVERY_PAYLOAD: discovery_payload,
+                    ATTR_DISCOVERY_TOPIC: topic,
+                }
+                setattr(discovery_payload, "discovery_data", discovery_data)
+
+            if discovery_hash in mqtt_data.discovery_pending_discovered:
+                pending = mqtt_data.discovery_pending_discovered[discovery_hash][
+                    "pending"
+                ]
```

Comment:
nit: this could be cleaned up to not access it every loop (probably doesn't matter though)
```diff
diff --git a/homeassistant/components/mqtt/discovery.py b/homeassistant/components/mqtt/discovery.py
index 48bd115e477..3e85e271c0d 100644
--- a/homeassistant/components/mqtt/discovery.py
+++ b/homeassistant/components/mqtt/discovery.py
@@ -386,6 +386,7 @@ async def async_start(  # noqa: C901
                 MqttComponentConfig(component, object_id, node_id, discovery_payload)
             )
 
+        discovery_pending_discovered = mqtt_data.discovery_pending_discovered
         for component_config in discovered_components:
             component = component_config.component
             node_id = component_config.node_id
@@ -416,10 +417,8 @@ async def async_start(  # noqa: C901
                 }
                 setattr(discovery_payload, "discovery_data", discovery_data)
 
-            if discovery_hash in mqtt_data.discovery_pending_discovered:
-                pending = mqtt_data.discovery_pending_discovered[discovery_hash][
-                    "pending"
-                ]
+            if discovery_hash in discovery_pending_discovered:
+                pending = discovery_pending_discovered[discovery_hash]["pending"]
                 pending.appendleft(discovery_payload)
                 _LOGGER.debug(
                     "Component has already been discovered: %s %s, queuing update",
```